planner/core/rule_join_reorder_greedy.go

```diff
@@ -0,0 +1,203 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+import (
+	"math"
+	"sort"
+
+	"github.com/pingcap/parser/ast"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/sessionctx"
+)
+
+func extractInnerJoinGroup(p *LogicalJoin) ([]LogicalPlan, []*expression.ScalarFunction, []expression.Expression) {
+	if p.reordered || p.preferJoinType > uint(0) || p.JoinType != InnerJoin || p.StraightJoin {
+		return nil, nil, nil
+	}
+	lChild := p.children[0]
+	rChild := p.children[1]
+	lJoin, ok := lChild.(*LogicalJoin)
+	if !ok {
+		eqConds := make([]*expression.ScalarFunction, len(p.EqualConditions))
+		otherConds := make([]expression.Expression, len(p.OtherConditions))
+		copy(eqConds, p.EqualConditions)
+		copy(otherConds, p.OtherConditions)
+		return []LogicalPlan{lChild, rChild}, eqConds, otherConds
+	}
+	lhsJoinGroup, eqEdges, otherConds := extractInnerJoinGroup(lJoin)
+	if lhsJoinGroup == nil {
+		return nil, nil, nil
+	}
+	return append(lhsJoinGroup, rChild), append(eqEdges, p.EqualConditions...), append(otherConds, p.OtherConditions...)
+}
+
+type joinReOrderGreedySolver struct {
+	ctx          sessionctx.Context
+	curJoinGroup []LogicalPlan
+	eqEdges      []*expression.ScalarFunction
+	otherConds   []expression.Expression
+}
+
+func (s *joinReOrderGreedySolver) solve() (LogicalPlan, error) {
+	for _, node := range s.curJoinGroup {
+		_, err := node.deriveStats()
+		if err != nil {
+			return nil, err
+		}
+	}
+	sort.Slice(s.curJoinGroup, func(i, j int) bool {
+		return s.curJoinGroup[i].statsInfo().RowCount < s.curJoinGroup[j].statsInfo().RowCount
```

Comment:
Seems `statsInfo().RowCount` of logical plan nodes does not reflect `ExpectedCnt` if `LIMIT` exists in the query, IIRC they are adjusted in physical optimization by `StatsInfo::ScaleByExpectCnt`, so this may lead to some inaccuracy, and it looks not easy to resolve.

Reply:
Yes, the `ExpectCnt is hard to estimate here since it needs to consider the actual physical implementation of this join node.